lstests/doc/README.md

```diff
@@ -0,0 +1,688 @@
+thread-2: first half
+thread-2: second half
+```
+
+or
+
+```
+thread-2: first half
+thread-2: second half
+thread-1: first half
+thread-1: second half
+```
+
+but will never be of the following form
+
+```
+thread-2: first half
+thread-1: first half
+thread-1: second half
+thread-2: second half
+```
+
+#### Log Levels
+
+| level   | value |
+| ------- | ----- |
+| `OFF`   | 100   |
+| `FLAG`  | 90    |
+| `FATAL` | 60    |
+| `ERROR` | 50    |
+| `WARN`  | 40    |
+| `INFO`  | 30    |
+| `DEBUG` | 20    |
+| `TRACE` | 10    |
+| `ALL`   | 0     |
+
+The `LogLevel` class accepts either a numeric value or a case-insensitive string for the level. While there are text names given for the level values, all the integers in `[0, 100]` are valid log levels.
+
+```python
+clus.log("Some warning information.", level="warn")
+```
+
+is equivalent to
+
+```python
+clus.log("Some warning information.", level=40)
+```
+
+or
+
+```python
+clus.warn("Some warning information.")
+```
+
+As another example,
+
+```python
+clus.log("Some info more prominent than INFO but not yet a WARN", level=35)
+```
+
+This line will be written to any logging destination with a threshold at or lower than 35. It will be printed out on the screen if the `ScreenWriter` has an `INFO`-level threshold. It will be written to a log file if there is a file writer `FileWriter(threshold=35)`. But it will not appear on the log file that captures only `WARN`-or-higher-level information.
+
+### `Service`
+
+A `Service` object represents the connection with the launcher service running in the background. Once a `Service` object is created, it will try to connect the launcher service automatically.
+
+After configuration, the `Service` will change the working directory, register (and possibly *modify*) the Logger, and then connect to launcher service. *Currently, it is only possible to connect to a local launcher service.*
+
+If there is already an existing launcher service running in the background, the `Service` object will by default connect to it (without starting a ew one). It is possible to override this default behavior by requesting to always start a new launcher service and/or to kill all the existing launcher service(s).
+
+A detailed explanation of the parameters to initialize a `Service` object can be found in its docstring.
+
+     Parameters
+     ----------
+    logger : Logger
+        Logger object which controls logging behavior.
+    addr : str
+        IP address of launcher service.
+        Currently, only local launcher service is supported. That is, only
+        default value "127.0.0.1" is supported.
+    port : int
+        Listening port of launcher service.
+        If there are multiple launcher service running in the background,
+        they must have different listening ports.
+        Default is 1234.
+    wdir : str
+        Working directory.
+        Default is the build folder.
+    file : str
+        Path to local launcher service file.
```

Comment:
What is a local launcher service file? How is the user supposed to use it. It would be helpful to document this.

Reply:
`file` is the path to the local executable file of the launcher service. It is `./programs/launcher-service/launcher-service` by default (relative path based on the `build` folder of `eos`). Normally users won't need to specify it. They only need to specify it if (1) there is no existing launcher service in the background, and (2) they want to start the launcher service via their python scripts via `core.service` (instead of manually starting launcher service from command line), and (3) they have a customized working directory so the executable path is not the default value.

Reply:
The change in description helps a lot. Thanks.

You said the relative path is based on the `build` folder of `eos`. But isn't that with the old behavior of the default working directory being the `build` folder. With the new change that makes the default working directory the current working directory, the relative paths would all be relative to the current working directory by default, right? And if the user specifies an alternative working directory using `wdir`, would the relative paths of other paths (like `file` or `gene`) be relative to that specified directory?

Reply:
Thanks for the questions. Let me clarify.

When I said "relative path based on the `build` folder of `eos`", I meant and only meant to explain the default value of `file`, which is `./programs/launcher-service/launcher-service`, is a relative path. That's it.

The new behavior makes no assumption about the existence of a `build` folder.

> With the new change that makes the default working directory the current working directory, the relative paths would all be relative to the current working directory by default, right? 

Yes, all the relative paths would all be relative to the current working directory by default.

> And if the user specifies an alternative working directory using `wdir`, would the relative paths of other paths (like `file` or `gene`) be relative to that specified directory?

Yes, if the user specifies an alternative working directory using `wdir`, the relative paths of other paths (like `file` or `gene`) would be relative to that specified directory.